src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp

```diff
@@ -1280,6 +1321,31 @@ std::weak_ptr<WaveClip> WaveTrackView::GetSelectedClip()
    return {};
 }
 
+std::vector<std::shared_ptr<WaveTrackSubView>> WaveTrackView::GetVisibleSubviews()
+{
```

Comment:
But this new function -- I don't see why it is needed.  I think you could solve this problem equally well with GetSubViews(), and it would be preferable to do so without a duplication of logic.  GetSubViews() reports some extra information you can ignore, which is not a significant extra expense.

If you convince me I'm wrong and that there is a real need for another function, then I'd just ask for a name change, up-casing the second V for consistency.

Reply:
Is there a way to get information about subview order using only GetSubViews()? There are couple of comments in hit testing where i tried to explain that we need to know if we are testing the top most cell (for those we don't want to hit test top area of the cell), the most bottom (for those we don't want to test the bottom area of the cell) or the cell between them (for future more-than-two-channels case we need both top and bottom tests)

Reply:
In the proposed method of cell hit testing I need to know whether tested subview is first or last. Hit testing area depends on that, that explained in the comments. What's your suggestion for proper hit testing?

Reply:
Yes, you need to know whether the subview is at the top or bottom.

There are already two existing functions already to describe subviews.

`GetAllSubViews()` is not right for your purpose, because the ordering ignores the display setting. 

But I think `GetSubViews()` is already right for your purposes and makes the new function unnecessary.

Reply:
Yep, sorry, i confused them. Yet, GetSubView changes the internal state of the WaveTrackView and requires to supply a rect of the track view cell, the height of which later is used by SubViewAdjustHandle, and the actual rect of the parent WaveTrackView is unknown during hit testing of the subview (if there is a good way to get from there then we can avoid subview order check at all), do still you think that it would be better to use it here?

Reply:
If you supply an arbitrary non-empty rectangle, you will get meaningful results about the ordering of sub-views.  You will also get some y-coordinates that you don't need, dividing the height of the rectangle in correct proportions.

Reply:
Calling GetSubView() with arbitrary rectangle will change the cached height of the WaveTrackView affecting the results of the behaviour of SubViewAdjustHandle, still I can call it like:
`auto subviews = pWaveTrackView->GetSubViews(wxRect(0, 0, 1, pWaveTrackView->GetLastHeight()));`
and make it public (it is declared public in base class actually)

Reply:
It is good that you noticed the state changes and asked questions.

One of the state changes is a "lazy evaluation" building the sub-views only on demand.  That should not be a problem.  

(The reason why it's needed is commented `// On-demand steps that can't happen in the constructor` and has to do with some difficulties around std::shared_from_this.)

The other is caching of mLastHeight, which might cause trouble for some other hit-testing routines.

I suggest:  make a version of GetSubViews taking a pointer to wxRect.  (Existing GetSubViews will call the new function, but needs to override TrackView::GetSubViews() which will not change.). When the pointer is null, supply a dummy rectangle instead and skip the cache update.

A little inelegant, but it will work and avoid duplication of logic.

Reply:
It's called side effects

Reply:
I admit there is some inelegancy, but this does eliminate duplication.  Thanks!